bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/CommerceRedirectServlet.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.sling.api.servlets.ServletResolverConstants;
+import org.apache.sling.api.servlets.SlingSafeMethodsServlet;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import com.adobe.cq.commerce.core.components.services.UrlProvider;
+import com.adobe.cq.commerce.core.components.services.UrlProvider.IdentifierLocation;
+import com.adobe.cq.commerce.core.components.services.UrlProvider.ProductIdentifierType;
+import com.adobe.cq.commerce.core.components.utils.SiteNavigation;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+import com.day.cq.wcm.api.PageManagerFactory;
+
+@Component(
+    service = Servlet.class,
+    immediate = true,
+    property = {
+        ServletResolverConstants.SLING_SERVLET_METHODS + "=" + HttpConstants.METHOD_GET,
+        ServletResolverConstants.SLING_SERVLET_RESOURCE_TYPES + "=" + CommerceRedirectServlet.RESOURCE_TYPE,
+        ServletResolverConstants.SLING_SERVLET_SELECTORS + "=" + CommerceRedirectServlet.PRODUCT_SELECTOR + ","
+            + CommerceRedirectServlet.CATEGORY_SELECTOR,
+        ServletResolverConstants.SLING_SERVLET_EXTENSIONS + "=" + CommerceRedirectServlet.EXTENSION
+    })
+public class CommerceRedirectServlet extends SlingSafeMethodsServlet {
+
+    protected static final String PRODUCT_SELECTOR = "cifproductredirect";
+    protected static final String CATEGORY_SELECTOR = "cifcategoryredirect";
+    protected static final String EXTENSION = "html";
+    protected static final String RESOURCE_TYPE = "core/cif/components/structure/page/v1/page";
+
+    private Pair<IdentifierLocation, ProductIdentifierType> productIdentifierConfig;
+    private Page productPage;
+    private Page categoryPage;
+
+    @Reference
+    private UrlProvider urlProvider;
+
+    @Reference
+    PageManagerFactory pageManagerFactory;
+
+    @Override
+    protected void doGet(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {
+        String suffix = request.getRequestPathInfo().getSuffix();
+        String[] selectors = request.getRequestPathInfo().getSelectors();
+
+        if (suffix == null) {
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Missing redirect suffix.");
+            return;
+        }
+
+        String[] suffixInfo = suffix.substring(1).split("/");
+        if (suffixInfo.length != 1) {
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Redirect suffix has wrong format.");
+            return;
+        }
+
+        PageManager pageManager = pageManagerFactory.getPageManager(request.getResourceResolver());
+
+        Page currentPage = pageManager.getContainingPage(request.getResource());
+        productPage = SiteNavigation.getProductPage(currentPage);
+        categoryPage = SiteNavigation.getCategoryPage(currentPage);
+
+        if (PRODUCT_SELECTOR.equals(selectors[0])) {
```

Comment:
Can you split the servlet into 2 servlets one doing the product selector the other doing the category selector? You can keep the common parts in an abstract class but this will keep the check on selector in Sling.